### PR TITLE
(PDB-4017) Accept and test rich data in terminus

### PR DESCRIPTION
**Pull request by austb**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -98,7 +104,7 @@ jobs:
       script: *run-integration-tests
 
     - stage: ❧ pdb tests
-      env: PDB_TEST=int/oraclejdk10/pup-master/srv-master/pg-9.6
+      env: PDB_TEST=int/rich/oraclejdk10/pup-master/srv-master/pg-9.6
```
Comment (rbrw) [inline .travis.yml]: Hmm, I'm not sure this is the set of configurations we want, but we can figure that out and adjust if needed later/soon.

File: test/puppetlabs/puppetdb/integration/reports.clj [hunk 2]
```diff
@@ -101,6 +101,86 @@
...
+     {:name "string_with_double_quote" :code "\"he said \\\"hi\\\"\"" :expected "he said \"hi\""}
+     {:name "regexp" :code "/[a-z]+/" :expected "/[a-z]+/"}
+     {:name "deferred" :code "Deferred('join', [[1, 2, 3], ':'])" :expected "1:2:3"}
+     {:name "sensitive_deferred" :code "Sensitive(Deferred('join', [[1, 2, 3], ':']))" :expected-old-value "[redacted]" :expected "[redacted]"}
+     {:name "sensitive" :code "Sensitive('password')" :expected-old-value "[redacted]" :expected "[redacted]"}
+     {:name "timestamp" :code "Timestamp('2012-10-10')" :expected "2012-10-10T00:00:00.000000000 UTC"}
+     {:name "hash_and_array" :code "{'a' => [1, 2, 3], 'b' => 'hello'}" :expected "{\"a\"=>[1, 2, 3], \"b\"=>\"hello\"}"}
+     {:name "special_key" :code "{ ['special', 'key'] => 10 }" :expected "{\"[\\\"special\\\", \\\"key\\\"]\"=>10}"}
+     {:name "hash_with_sensitive_val" :code "{ x => Sensitive(hush) }" :expected #"\{\"x\"=>\"#<Sensitive \[value redacted\]:[0-9]+>\"}"}
+     {:name "hash_with_sensitive_key" :code "{Sensitive(hush) => 42 }" :expected #"\{\"#<Sensitive \[value redacted\]:[0-9]+>\"=>42\}"}
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/reports.clj]: Perhaps make these longer ones at least two lines so they're easily visible in gh, etc.

File: test/puppetlabs/puppetdb/integration/reports.clj [hunk 3]
```diff
@@ -101,6 +101,86 @@
...
+     {:name "literal_integer" :code "47" :expected "47"}
+     {:name "literal_float" :code "3.14" :expected "3.14"}
+     {:name "literal_true" :code "true" :expected "true"}
+     {:name "literal_string" :code "\"hello\"" :expected "hello"}
+     {:name "string_with_single_quote" :code "\"ta'phoenix\"" :expected "ta'phoenix"}
+     {:name "string_with_double_quote" :code "\"he said \\\"hi\\\"\"" :expected "he said \"hi\""}
+     {:name "regexp" :code "/[a-z]+/" :expected "/[a-z]+/"}
+     {:name "deferred" :code "Deferred('join', [[1, 2, 3], ':'])" :expected "1:2:3"}
+     {:name "sensitive_deferred" :code "Sensitive(Deferred('join', [[1, 2, 3], ':']))" :expected-old-value "[redacted]" :expected "[redacted]"}
+     {:name "sensitive" :code "Sensitive('password')" :expected-old-value "[redacted]" :expected "[redacted]"}
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/reports.clj]: Maybe change :expected-old-value to :expected-old to match :expected?

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 4]
```diff
@@ -27,3 +28,94 @@
           (is (= catalog-uuid report-uuid)))))))
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: Is the link to the friendly format doc (examples, etc.) "stable"?  If so, let's put it in a comment here.

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 5]
```diff
@@ -27,3 +28,94 @@
...
+                ps (int/run-puppet-server [pdb] {:agent {:rich-data true}})]
+
+      (testing "Agent run succeeds"
+        (let [manifest (str "type Car = Object[attributes => {regnbr => String}];\n"
+                            (apply str (for [{:keys [name code]} rich-data-tests]
+                                         (str "notify { '" name "':"
+                                              "  message => " code
+                                              "}"))))
+              {:keys [out]} (int/run-puppet-as "my-agent" ps pdb manifest {:rich-data true})]
+          (is (re-find #"a-z" out))))
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: This is actually looking for the three literal characters in sequence "a-z"?

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 6]
```diff
@@ -27,3 +28,94 @@
...
+
+      (testing "Agent run data can be queried"
+        (are [query result] (= result (int/pql-query pdb query))
+          "nodes[certname] {}" [{:certname "my-agent"}]
+          "catalogs[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "factsets[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "reports[certname, environment] {}" [{:certname "my-agent", :environment "production"}]))
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: If you prefer this, no worries, but these days I'm tempted to use "is" instead of "are" even if it's a little chattier so that the line numbers will be right when something goes wrong.

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 7]
```diff
@@ -27,3 +28,94 @@
...
+
+      (testing "Catalog resource parameters stored in proper format"
+        (let [[{{all-resources :data} :resources} :as result] (int/pql-query pdb "catalogs[certname, resources] { resources { type = 'Notify' and title = 'regexp' } }")
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: Let's wrap this, perhaps either with some additional (intermediate) let bindings, or some threading forms, or even just

    (let [[...........something long............]
           (some-long-call ...)

           next-binding ...

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 8]
```diff
@@ -27,3 +28,94 @@
...
+              filtered-resources (filter
+                                   #(not (or (= "Stage" (:type %))
+                                             (and (= "Class" (:type %))
+                                                  (or (= "main" (:title %))
+                                                      (= "Settings" (:title %))))))
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: I'd need to think about the logic for a minute, but could this be a bit shorter as a remove?  i.e. (remove #(or ...

Comment (rbrw) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: Also, not critical, but I'd be tempted to just shorten this to "resources", since it's the resources of interest, the scope is localized, and we already have all-resources as a comparative term.

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 9]
```diff
@@ -27,3 +28,94 @@
...
+              (if (:valid? test-resource)
+                (is ((:valid? test-resource) stored-value))
+                (is (= (:expected test-resource) stored-value)))))))
+
+      (testing "transaction-uuid"
+        (let [catalog-uuid (first (int/pql-query pdb "catalogs [transaction_uuid] {}"))
+              report-uuid (first (int/pql-query pdb "reports [transaction_uuid] {}"))]
+          (testing "is available from puppetdb"
+            (is (not (nil? catalog-uuid)))
+            (is (not (nil? report-uuid))))
```
Comment (rbrw) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: Are we specifically looking for `nil` if so, then this is just right, but if not, perhaps change this to just `(is (not catalog-uuid)), etc.

File: test/puppetlabs/puppetdb/integration/smoke.clj (hunk 6 above)
Comment (austb) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: That one is straight copy-pasta from the above test, so I'm tempted to leave it unless you want to change both

File: test/puppetlabs/puppetdb/integration/smoke.clj (hunk 9 above)
Comment (austb) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: this is another one taken from above, so I'm tempted to leave it as is

File: test/puppetlabs/puppetdb/integration/smoke.clj (hunk 5 above)
Comment (austb) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: Ya, that's left over from when the test only used Regex, so the notify was "/[a-z]+/" and that made sure that it had been notified to the stdout of the puppet run

File: test/puppetlabs/puppetdb/integration/smoke.clj [hunk 10]
```diff
@@ -27,3 +28,94 @@
           (is (= catalog-uuid report-uuid)))))))
 
 
+(when test-rich-data?
+  (def rich-data-tests
```
Comment (Zak-Kent) [inline test/puppetlabs/puppetdb/integration/smoke.clj]: Are we holding off on testing the alien objects for the time being? I just noticed them in the doc but I'm not sure what they are exactly and if we need to be looking out for them.

Comment (austb) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: Yeah we are waiting on that (but I need to open a ticket to add those), they would require also loading some ruby code to bootstrap those objects, which we don't have the infrastructure for in place right now

File: test/puppetlabs/puppetdb/integration/smoke.clj (hunk 4 above)
Comment (austb) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: I'm hesistant to link to the doc because it's inaccurate, but I can create a new one that's accurate and link it here if you want

Comment (rbrw) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: Oh, I meant *their* doc.  If puppet doesn't have an official one yet for this format, then we can wait, but we should eventually have something to refer to.

File: test/puppetlabs/puppetdb/integration/smoke.clj (hunk 6 above)
Comment (rbrw) [reply test/puppetlabs/puppetdb/integration/smoke.clj]: Not critical either way.
